File: ios/ParameterTypes.swift
//
//  ParameterTypes.swift
//  ExpoPhotoManipulator
//
//  Created by Woraphot Chokratanasombat on 9/8/2567 BE.
//

import CoreGraphics
import ExpoModulesCore
import WCPhotoManipulator

internal struct Point: Record {
    @Field
    var x: Double

    @Field
    var y: Double
    
    func toPoint() -> CGPoint {
        return CGPointMake(x, y)
    }
    
    static func from (_ x: Double, _ y: Double) -> Point {
        let result = Point()
        result.x = x
        result.y = y
        return result
    }
}

internal struct Size: Record {
    @Field
    var width: Double

    @Field
    var height: Double
    
    func toCGSize() -> CGSize {
        return CGSizeMake(width, height)
    }
}

internal struct Color: Record {
    /***
     * Red
     */
    @Field
    var r: Double
    
    /***
     * Green
     */
    @Field
    var g: Double
    
    /***
     * Blue
     */
    @Field
    var b: Double
    
    /***
     * Alpha
     */
    @Field
    var a: Double
    
    func toUIColor() -> UIColor {
        return UIColor(red: component(r), green: component(g), blue: component(b), alpha: component(a))
    }
    
    private func component(_ value: Double) -> Double {
        return value / 255
    }
    
    static func from (_ r: Double, _ g: Double, _ b: Double, _ a: Double) -> Color {
        let result = Color()
        result.r = r
        result.g = g
        result.b = b
        result.a = a
        return result
    }
}

internal struct Rect: Record {
    @Field
    var x: Double

    @Field
    var y: Double
    
    @Field
    var width: Double

    @Field
    var height: Double
    
    func toCGRect() -> CGRect {
        return CGRectMake(x, y, width, height)
    }
}

internal struct TextOptions: Record {
    @Field
    var position: Point
    
    @Field
    var text: String
    
    @Field
    var textSize: Double
    
    @Field
    var fontName: String?
    
    @Field
    var color: Color
    
    @Field
    var thickness: Double
    
    @Field
    var rotation: Double
    
    @Field
    var shadowRadius: Double

    @Field
    var shadowOffset: Point?
    
    @Field
    var shadowColor: Color?
    
    func getPosition() -> CGPoint {
        return position.toPoint()
    }
    
    func getTextStyle() -> TextStyle {
        return TextStyle(
                color: color.toUIColor(),
                font: ParamUtils.font(fontName, size: textSize),
                thickness: thickness,
                rotation: rotation,
                shadowRadius: shadowRadius,
                shadowOffsetX: toInt(shadowOffset?.x, 0),
                shadowOffsetY: toInt(shadowOffset?.y, 0),
                shadowColor: shadowColor?.toUIColor()
        )
    }
    
    private func toInt(_ value: Double?, _ defaultValue: Int) -> Int {
        return value.map { Int($0) } ?? defaultValue
    }
}

File: android/src/main/java/com/guhungry/expophotomanipulator/ParameterTypes.kt
package com.guhungry.expophotomanipulator

import android.graphics.Color as AColor
import android.graphics.PointF
import com.guhungry.photomanipulator.model.CGRect
import com.guhungry.photomanipulator.model.CGSize
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class Point(
    @Field val x: Float = 0f,
    @Field val y: Float = 0f
): Record {
    fun toPointF(): PointF {
        return PointF(x, y)
    }
}

class Size: Record {
    @Field
    val width: Int = 0

    @Field
    val height: Int = 0

    fun toCGSize(): CGSize {
        return CGSize(width, height)
    }
}

class Color(
    // Red
    @Field val r: Int = 0,

    // Green
    @Field val g: Int = 0,

    // Blue
    @Field val b: Int = 0,
        // Alpha
    @Field val a: Int = 0
): Record {
    fun toColor() = AColor.argb(a, r, g, b)
}

class Rect: Record {
    @Field
    val x: Int = 0

    @Field
    val y: Int = 0

    @Field
    val width: Int = 0

    @Field
    val height: Int = 0

    fun toCGRect(): CGRect {
        return CGRect(x, y, width, height)
    }

}

class TextOptions(
    @Field
    val position: Point = Point(),

    @Field
    val text: String = "",

    @Field
    val textSize: Float = 0f,

    @Field
    val fontName: String? = null,

    @Field
    val color: Color = Color(),

    @Field
    val thickness: Float = 0f,

    @Field
    val rotation: Float = 0f,

    @Field
    val shadowRadius: Float = 0f,

    @Field
    val shadowOffset: Point? = null,

    @Field
    val shadowColor: Color? = null
) : Record